add headerTags to pageBanner

labbydev opened the pull request:
Include the headerTags as options to the pageBanner

New patterns:
* Content Eyebrow organism (http://localhost:3000/?p=organisms-content-eyebrow)

legostud commented on styleguide/source/_patterns/03-organisms/by-template/page-banner.twig:
```diff
@@ -2,6 +2,16 @@
 {% set colorClass = pageBanner.color ? "ma__page-banner--" ~ pageBanner.color : '' %}
 
+{% if pageBanner.headerTags.taxonomyTerms %}
+  <div class="ma__page-banner__tags">
```
This is an anti-pattern since you can't have a child `ma__page-banner__tags` element located outside of a parent element `ma__page-banner`.

jesconstantine replied:
I would suggest moving this conditional block inside of the parent component `section.ma__page-banner` similar to how we see [@molecules/header-tags included on @organisms/by-template/page-header](https://github.com/massgov/mayflower/blob/master/styleguide/source/_patterns/03-organisms/by-template/page-header.twig#L6).  What do you think @labbydev @legostud ?

legostud replied:
It can't be added there, because the background image is applied to the parent class so we would have to do a lot more work to tweak the Page Banner.  I think making it a stand alone pattern placed above the Page Banner would be a better approach.

jesconstantine replied:
Good call @legostud that makes sense to me.

legostud commented on styleguide/source/assets/scss/03-organisms/_page-banner.scss:
```diff
@@ -323,4 +323,29 @@ $page-banner-icon-width: $column + $gutter;
     }
   }
+
+  &__tags {
```
I think you forgot to also create the 'theme' version (colors, and font-weights) of this code.

jesconstantine commented:
@legostud @gleroux02 let's talk about the architecture needed for this ticket and what is proposed in Jonathan's review comment so we can provide a path forward for @labbydev 

It seemed logical to me to add the `@molecules/header-tags` as a child pattern to whichever `@organism` was responsible for rendering the page title and other "pre-content" of the page, which would mean adding it to `@organisms/by-template/page-banner` `@organisms/by-template/illustrated-header` as it already exists on `@organisms/by-template/page-header`.  

The reason why it seemed logical to me is because in order to render, this pattern needs page-level data (i.e. relationship information--not hierarchical) and not site-level data.

What do people think?

legostud commented:
if it needs page-level data, then it might not be best to add it to the header.  I would then suggest we make it a stand alone Pattern that can be placed at the top of the page above the various Page Headers/Banners.  Similar to how we were doing breadcrumbs.

legostud commented on styleguide/source/_patterns/03-organisms/by-template/header-tags.json:
```diff
@@ -0,0 +1,37 @@
+      "linkType": "linkedin"
+    }]
+  }
+}
```
Channeling Greg here, you're missing the extra line ending on your new files.

legostud commented on styleguide/source/_patterns/03-organisms/by-template/header-tags.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "headerTags": {
```
The data object for all Patterns in Mayflower, excluding pages and templates, should be contained in a single object that matches the files name.  This of course means your going to have { headerTags: { headerTags: { } } } which tells me this new pattern is named incorrectly.

Example: sample-pattern.json
~~~
{
  samplePattern {
    headerTags { ... }
    socialLinks { ... }
  }
}
~~~

legostud commented on styleguide/source/_patterns/03-organisms/by-template/header-tags.twig:
```diff
@@ -0,0 +1,13 @@
+
+{% if headerTags.taxonomyTerms or socialLinks.items %}
+  <div class="ma__header-tags__container">
```
Incorrect BEM syntax used here.  I see a child class that isn't within a parent class.

legostud commented on styleguide/source/assets/scss/03-organisms/_header-tags.scss:
```diff
@@ -0,0 +1,33 @@
+.ma__header-tags__container {
```
Something is obviously wrong if the outer wrapper is a child class.  Also since this pattern is called header-tags, the parent class should be .ma__header-tags { ... }.  Filename determines the class name to keep everything in sync and help avoid having multiple patterns using to same parent class.

Using .ma__header-tags isn't possible here since we are already using that elsewhere.

legostud commented on styleguide/source/assets/scss/03-organisms/_header-tags.scss:
```diff
@@ -0,0 +1,33 @@
+.ma__header-tags__container {
+  @include ma-container;
+  clear: both;
```
I don't think we'll need the clear both here.

legostud commented on styleguide/source/assets/scss/index.scss:
```diff
@@ -119,6 +119,7 @@
 @import "03-organisms/form-requirements";
 @import "03-organisms/header";
 @import "03-organisms/header-alert";
+@import "03-organisms/header-tags";
```
I'm pretty sure there are theme SCSS styles (border-color) for this pattern.

labbydev replied:
@legostud I couldn't find where the border color for this is being defined, do you know where that is?

labbydev commented:
@legostud @jesconstantine take a look at this update and LMK how it looks to you.

jesconstantine commented:
@legostud @labbydev Does it make sense to call the pattern something like Content Eyebrow?

So `@organisms/by-author/content-eyebrow` would include `@molecules/header-tags` and `@molecules/social-links`.

Data structure would be:
```
{
  contentEyebrow: {
    headerTags: {},
    socialLinks: {}
  }
}
```

And we'd still remove `headerTags` from `pageHeader` -- can we do this in a minor version?

legostud commented:
Creating a new pattern = minor release.  
Removing code from the twig and data object for Page Header = minor release.  
Removing the CSS from page-header.scss = Major release (breaks pre-existing markup).

A minor release needs to work with pre-existing markup and data from the current Major release (5).  Adding the new pattern is easy, but modifying the Page Header will take some care to keep it backward compatible.  Technically, we don't need to modify page-header at all since the header tags are optional and can just be turned off by not passing any data.  I would still request it be removed, but that work could be done as part of a future 6.0 release.  

My suggestion is to do two PRs.  One to create the new pattern and update the pages in Mayflower (minor release) and one to remove the header tags and social links from all the Page Header files (major release).

jesconstantine commented:
Assuming @labbydev creates the PR that you mention (creating the new pattern) for a minor release for this ticket; @labbydev can implement that new pattern on pages that need the `headerTags` but currently use the `pageBanner` and leave the pages that use the `pageHeader` to send `headerTags` as is until we implement the future PR / major release to remove the `headerTags` + `socialLinks` from `pageHeader` and updates those pages that include `pageHeader` to use the new pattern.  I think that's out of scope for this ticket for Lauren, but Iet's at least create the ticket and ear mark it for future refactoring.

How does that sound for a path forward for everyone?

legostud commented:
sounds fine with me.

labbydev commented:
@jesconstantine @legostud sounds good to me, I just want to confirm the action items:
* [ ]  Create new ticket to remove `headerTags` from `pageHeader` 
   _@jesconstantine is this something you're going to do or do you want me to do this?_
* [ ] Create a new pattern for the `headerTags` 
   _`contentEyebrow` sounds promising_
* [ ] Put the styles back in `page-header.scss` to prevent backward breaking changes
* [ ] If a pattern is already using `pageHeader` leave it as is (with `headerTags` and `socialLinks` included)
* [ ] Additional scss fixes pointed out by @legostud (see previous comments)

Does this seem correct?

jesconstantine commented:
@labbydev This sounds right to me as well.  I'll create the ticket mentioned in your first bullet.

labbydev commented:
Thanks, sounds like a plan

labbydev commented:
@jesconstantine I have added the new organism and reverted the breaking changes with `pageHeader` would you like to me to add the `contentEyebrow` to any of the other patterns?

legostud commented on styleguide/source/_patterns/03-organisms/by-template/content-eyebrow.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "contentEyebrow": {
```
perfect

jesconstantine replied:
@legostud based on [our discussion re: `by-template` vs `by-author`](https://github.com/massgov/mayflower/pull/527#discussion_r131629182), should this pattern live under `by-author` instead?

legostud replied:
If it was in the by-author area I would expect the user to be able to place this pattern anywhere on the page (in one of the various placeholders like Page Content or Right Rail).  This new pattern is only designed to be use in one location (Pre-Conntent) and should be at the top of that location so it has very strict placement requirements.

legostud commented on styleguide/source/_patterns/03-organisms/by-template/content-eyebrow.twig:
```diff
@@ -0,0 +1,14 @@
+<section class="ma__content-eyebrow">
+  {% if contentEyebrow.headerTags.taxonomyTerms or contentEyebrow.socialLinks.items %}
```
I would put this outside of the section tag to remove all the html if there aren't any items to render.

legostud commented on styleguide/source/_patterns/03-organisms/by-template/content-eyebrow.twig:
```diff
@@ -0,0 +1,14 @@
+      {% endif %}
+    </div>
+  {% endif %}
+</section>
```
line ending.

labbydev commented:
@legostud where is that screenshot coming from? I don't see that code in `styleguide/source/assets/scss/03-organisms/_page-header.scss`? Could you possibly link me to the line in github? I'm confused since the code is basically the same between `_page-header.scss` and `_content-eyebrow.scss` what I'm missing when it comes to the border color.

labbydev commented:
@legostud @jesconstantine RE:Illustrated header not getting the content eyebrow, since guide pages use the illustrated header, but also require relationship indicators, how should that be addressed?

labbydev commented:
Should we have an option for border or no border? Seems awkward to see the border with the banner.
<img width="1185" alt="screen shot 2017-08-14 at 1 26 47 pm" src="https://user-images.githubusercontent.com/1653723/29285698-f901f4e8-80f4-11e7-89b8-942016fa8ed3.png">

jesconstantine commented:
@labbydev @legostud for this PR let's stick with the list of page types from my [earlier comment](https://github.com/massgov/mayflower/pull/539#pullrequestreview-55965628) -- with the exception of the org page, the majority of the other page types seem to have them already anyway.

legostud commented on styleguide/source/_patterns/03-organisms/by-template/content-eyebrow.twig:
```diff
@@ -1,5 +1,5 @@
 {% if contentEyebrow.headerTags.taxonomyTerms or contentEyebrow.socialLinks.items %}
-  <section class="ma__content-eyebrow">
+  <section class="ma__content-eyebrow{{contentEyebrow.hideBorder ? '--borderless' : ''}}">
```
modifier classes `--something` are always pair with a parent class when using BEM syntax.  This way the modifier only contains the CSS need to override or change the default styling (no duplication)

`ma__content-eyebrow ma__content-eyebrow--borderless`

legostud commented on styleguide/source/assets/scss/03-organisms/_content-eyebrow.scss:
```diff
@@ -35,4 +35,14 @@
+
+  &--borderless {
+    @include ma-container;
+    @include clearfix;
```
these two includes shouldn't be needed once the you fix my other comment.

legostud replied:
it will also need to be `&--borderless &__tags {` to achieve the proper specificity

rbayliss reviewed (approved):
Looks good to me!
